packages/flutter/lib/src/widgets/basic.dart

```diff
@@ -1554,15 +1583,107 @@ class ConstrainedBox extends SingleChildRenderObjectWidget {
+/// on an infinite canvas with no constraints. This container will then expand
+/// as much as it can within its own constraints and align the child based on
+/// [alignment].  If the container cannot expand enough to accommodate the
+/// entire child, the child will be clipped.
+///
+/// In debug mode, if the child overflows the container, a warning will be
+/// printed on the console, and black and yellow striped areas will appear where
+/// the overflow occurs.
+///
+/// See also:
+///
+///  * [ConstrainedBox] for a box which imposes constraints on its child.
+///  * [Container], a convenience widget that combines common painting,
+///    positioning, and sizing widgets.
+///  * [OverflowBox], a widget that imposes different constraints on its child
+///    than it gets from its parent, possibly allowing the child to overflow
+///    the parent.
+///  * [RenderUnconstrainedBox] for the [RenderObject] for this container.
+///  * The [catalog of layout containers](https://flutter.io/widgets/layout/).
+class UnconstrainedBox extends SingleChildRenderObjectWidget {
+  /// Creates a widget that imposes no constraints on its child, allowing it to
+  /// render at its "natural" size. If the child overflows the parents
+  /// constraints, a warning will be given in debug mode.
+  const UnconstrainedBox({
+    Key key,
+    Widget child,
+    this.textDirection,
+    this.alignment = Alignment.center,
+  }) : assert(alignment != null),
+       super(key: key, child: child);
+
+  /// The text direction to use when interpreting the [alignment] if it is an
+  /// [AlignmentDirectional].
+  final TextDirection textDirection;
+
+  /// The alignment to use when laying out the child.
+  ///
+  /// If this is an [AlignmentDirectional], then [textDirection] must not be
+  /// null.
+  ///
+  /// See also:
+  ///  * [Alignment] for non-[Directionality]-aware alignments.
+  ///  * [AlignmentDirectional] for [Directionality]-aware alignments.
+  final AlignmentGeometry alignment;
+
+  bool get _needTextDirection => alignment is AlignmentDirectional
+      && alignment != AlignmentDirectional.center;
+
+  /// The value to pass to [RenderUnconstrainedBox.textDirection].
+  ///
+  /// This value is derived from the [textDirection] property and the ambient
+  /// [Directionality]. The value is null if there is no need to specify the
+  /// text direction. In practice, there's always a need to specify the direction
+  /// except for when the alignment is [AlignmentDirectional.center].
+  ///
+  /// This method exists so that subclasses of [UnconstrainedBox] that create
+  /// their own render objects that are derived from [RenderUnconstrainedBox]
+  /// can do so and still use the logic for providing a text direction only when
+  /// it is necessary.
+  @protected
+  TextDirection getEffectiveTextDirection(BuildContext context) {
+    return textDirection ?? (_needTextDirection ? Directionality.of(context) : null);
+  }
```

Comment:
you probably don't need this, you can just inline it into updateRenderObject.

Reply:
Done.